internal/metamorph/processor_helpers.go

```diff
@@ -22,64 +23,86 @@ func (p *Processor) GetProcessorMapSize() int {
 }
 
-func updateStatusMap(statusUpdatesMap map[chainhash.Hash]store.UpdateStatus, statusUpdate store.UpdateStatus) {
-	foundStatusUpdate, found := statusUpdatesMap[statusUpdate.Hash]
+func (p *Processor) updateStatusMap(statusUpdate store.UpdateStatus) error {
+	currentStatusUpdate, err := p.getTransactionStatus(statusUpdate.Hash)
+	if err != nil {
+		if errors.Is(err, cache.ErrCacheNotFound) {
+			// if record doesn't exist, save new one
+			return p.setTransactionStatus(statusUpdate)
+		}
+		return err
+	}
 
-	if !found || shouldUpdateStatus(statusUpdate, foundStatusUpdate) {
+	if shouldUpdateStatus(statusUpdate, *currentStatusUpdate) {
 		if len(statusUpdate.CompetingTxs) > 0 {
-			statusUpdate.CompetingTxs = mergeUnique(statusUpdate.CompetingTxs, foundStatusUpdate.CompetingTxs)
+			statusUpdate.CompetingTxs = mergeUnique(statusUpdate.CompetingTxs, currentStatusUpdate.CompetingTxs)
 		}
-
-		statusUpdatesMap[statusUpdate.Hash] = statusUpdate
+		// TODO: combine status history
 	}
-}
 
-func (p *Processor) updateStatusMap(statusUpdate store.UpdateStatus) (map[chainhash.Hash]store.UpdateStatus, error) { //nolint:unused
-	statusUpdatesMap := p.getStatusUpdateMap()
+	return p.setTransactionStatus(statusUpdate)
+}
 
-	foundStatusUpdate, found := statusUpdatesMap[statusUpdate.Hash]
+func (p *Processor) setTransactionStatus(status store.UpdateStatus) error {
+	bytes, err := json.Marshal(status)
+	if err != nil {
+		return errors.Join(ErrFailedToSerialize, err)
+	}
 
-	if !found || shouldUpdateStatus(statusUpdate, foundStatusUpdate) {
-		if len(statusUpdate.CompetingTxs) > 0 {
-			statusUpdate.CompetingTxs = mergeUnique(statusUpdate.CompetingTxs, foundStatusUpdate.CompetingTxs)
-		}
+	err = p.cacheStore.MapSet(CacheStatusUpdateHash, status.Hash.String(), bytes)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-		statusUpdatesMap[statusUpdate.Hash] = statusUpdate
+func (p *Processor) getTransactionStatus(hash chainhash.Hash) (*store.UpdateStatus, error) {
+	bytes, err := p.cacheStore.MapGet(CacheStatusUpdateHash, hash.String())
+	if err != nil {
+		return nil, err
 	}
 
-	err := p.setStatusUpdateMap(statusUpdatesMap)
+	var status store.UpdateStatus
+	err = json.Unmarshal(bytes, &status)
 	if err != nil {
 		return nil, err
 	}
 
-	return statusUpdatesMap, nil
+	return &status, nil
 }
 
-func (p *Processor) setStatusUpdateMap(statusUpdatesMap map[chainhash.Hash]store.UpdateStatus) error { //nolint:unused
-	bytes, err := serializeStatusMap(statusUpdatesMap)
+func (p *Processor) getAndDeleteAllTransactionStatuses() (StatusUpdateMap, error) {
+	statuses := make(StatusUpdateMap)
+	keys, err := p.cacheStore.MapExtractAll(CacheStatusUpdateHash)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	err = p.cacheStore.Set(CacheStatusUpdateKey, bytes, processStatusUpdatesIntervalDefault)
-	if err != nil {
-		return err
+	for key, value := range keys {
+		hash, err := chainhash.NewHashFromStr(key)
+		if err != nil {
+			return nil, err
```

Comment:
I think we should log the error here, including all the data needed for manual incident handling, and continue execution